src/proofs/weighted_inner_product.rs

```diff
@@ -0,0 +1,915 @@
+    let mut next_pow: BigInt = y.clone();
+    let mut pow_y: Vec<BigInt> = Vec::with_capacity(n);
+    let order = FE::q();
+    for i in 0..n {
+        pow_y.push(next_pow);
+        next_pow = BigInt::mod_mul(&pow_y[i], &y, &order);
+    }
+    return pow_y;
+}
+
+fn weighted_inner_product(a: &[BigInt], b: &[BigInt], y: BigInt) -> BigInt {
+    assert_eq!(
+        a.len(),
+        b.len(),
+        "weighted_inner_product(a,b): lengths of vectors do not match"
+    );
+    let order = FE::q();
+    let y_powers = scalar_powers(y, a.len());
+    let n = a.len();
+    let out = (0..n)
+        .map(|i| {
+            let aibi = BigInt::mod_mul(&a[i], &b[i], &order);
+            BigInt::mod_mul(&aibi, &y_powers[i], &order)
+        })
+        .fold(BigInt::zero(), |acc, x| BigInt::mod_add(&acc, &x, &order));
+
+    return out;
+}
+
+#[cfg(test)]
+mod tests {
+    use curv::arithmetic::traits::{Converter, Modulo};
+    use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+    use curv::cryptographic_primitives::hashing::hash_sha512::HSha512;
+    use curv::cryptographic_primitives::hashing::traits::*;
+    use curv::elliptic::curves::traits::*;
+    use curv::BigInt;
+    use curv::{FE, GE};
+    use proofs::range_proof::generate_random_point;
+    use proofs::weighted_inner_product::WeightedInnerProdArg;
+    use proofs::weighted_inner_product::{scalar_powers, weighted_inner_product};
+
+    fn test_helper(n: usize) {
+        let KZen: &[u8] = &[75, 90, 101, 110];
+        let kzen_label = BigInt::from(KZen);
+
+        let g_vec = (0..n)
+            .map(|i| {
+                let kzen_label_i = BigInt::from(i as u32) + &kzen_label;
+                let hash_i = HSha512::create_hash(&[&kzen_label_i]);
+                generate_random_point(&Converter::to_vec(&hash_i))
+            })
+            .collect::<Vec<GE>>();
+
+        // can run in parallel to g_vec:
+        let h_vec = (0..n)
+            .map(|i| {
+                let kzen_label_j = BigInt::from(n as u32) + BigInt::from(i as u32) + &kzen_label;
+                let hash_j = HSha512::create_hash(&[&kzen_label_j]);
+                generate_random_point(&Converter::to_vec(&hash_j))
+            })
+            .collect::<Vec<GE>>();
+
+        let label = BigInt::from(2);
+        let hash = HSha512::create_hash(&[&label]);
+        let g = generate_random_point(&Converter::to_vec(&hash));
+        let label = BigInt::from(3);
+        let hash = HSha512::create_hash(&[&label]);
+        let h = generate_random_point(&Converter::to_vec(&hash));
+
+        let a: Vec<_> = (0..n)
+            .map(|_| {
+                let rand: FE = ECScalar::new_random();
+                rand.to_big_int()
+            })
+            .collect();
+
+        let b: Vec<_> = (0..n)
+            .map(|_| {
+                let rand: FE = ECScalar::new_random();
+                rand.to_big_int()
+            })
+            .collect();
+
+        let y_scalar: BigInt =
```

Comment:
you are also generating y_scalar inside the prover and verifier. 
The method to generate y is I believe " nothing up my sleeve" - we can take y to be hash of the group generator for example.

Reply:
Added `y` as an input of wip and generate `y` using NUMS method in range proof.